model_zoo/jag_utils/select_samples.cpp

```diff
@@ -81,209 +221,370 @@ int main(int argc, char **argv) {
+    int local_valid_index = 0;
+    for (int local_index=0; local_index<total; local_index++) {
+      if (index_map_exclude[fn].find(local_index) == index_map_exclude[fn].end()) {
+        int global_idx = local_valid_index+first;
+        ++local_valid_index;
+        if (random_indices.find(global_idx) != random_indices.end()) {
+          index_map_keep[fn].insert(local_index);
+          index_map_exclude[fn].insert(local_index);
         }
-        index_map_exclude[fn].insert(string_to_index[sample_id]);
-      }
-      if (index_map_exclude[fn].size() != bad) {
-        err << "exclude.size(): " << index_map_exclude[fn].size() << " should be: " << bad << " but isn't\n";
-        LBANN_ERROR(err.str());
       }
+    }
+    first += good;
+  }
+  cout << "FINISHED selecting samples based on random indices; time: " << lbann::get_time() - tm3 << endl;
 
-      int local_valid_index = 0;
-      for (int local_index=0; local_index<total; local_index++) {
-        if (index_map_exclude[fn].find(local_index) == index_map_exclude[fn].end()) {
-          int global_idx = local_valid_index+first;
-          ++local_valid_index;
-          if (random_indices.find(global_idx) != random_indices.end()) {
-            index_map_keep[fn].insert(local_index);
-            index_map_exclude[fn].insert(local_index);
-          }
-        }
+  if (index_map_exclude.size() != index_map_keep.size()) {
+    LBANN_ERROR("index_map_exclude.size() != index_map_keep.size()");
+  }
+  cout << "  FINISHED build_index_maps; time: " << lbann::get_time() - tm1 << endl;
+}
+
+void sanity_test_request() {
+  const string index_fn = options::get()->get_string("index_fn").c_str();
+  ifstream in(index_fn.c_str());
+  if (!in) {
+    LBANN_ERROR("failed to open ", index_fn, " for reading");
+  }
+
+  string line;
+  getline(in, line);
+  if (line != "CONDUIT_HDF5_EXCLUSION") {
+    LBANN_ERROR("error: 1st line in index file must contain: CONDUIT_HDF5_EXCLUSION\n");
+  }
+
+  int num_valid, num_invalid, num_files;
+  in >> num_valid >> num_invalid >> num_files;
+  int samples_per_list = options::get()->get_int("num_samples_per_list");
+  int num_lists = options::get()->get_int("num_lists");
+  int num_samples = samples_per_list * num_lists;
+  if (num_samples > num_valid) {
+    LBANN_ERROR("you requested a total of ", num_samples, " samples, but only ", num_valid, " are available");
+  }
+}
+
+void divide_selected_samples(
+    const unordered_map<string, unordered_set<int>> &index_map_keep,
+    vector<unordered_map<string, unordered_set<int>>> &sets) {
+  size_t samples_per_list = options::get()->get_int("num_samples_per_list");
+  size_t which = 0;
+  size_t count = 0;
+  size_t total = 0;
+  for (auto &it : index_map_keep) {
+    const string &filename = it.first;
+    const unordered_set<int> &sample_ids = it.second;
+    for (auto &it2 : sample_ids) {
+      sets[which][filename].insert(it2);
```

Comment:
Looking at this, I believe that it will interleave all of the samples across each training list being generated.  Is that correct?  If so, this is a step in the right direction, but we really should randomly pull a sample and then randomly assign it to the list.

Reply:
Although, given that it is pulling from a unordered set, maybe it is okay to interleave the assignment.

Reply:
Hm, I'll have to think on this (possibly very early tomorrow). 
"randomly pull a sample" - I believe the code does this
"and then randomly assign it to the list" - but I think the ++which effectively does this. OK, it's not random, but I think (??) should have the same effect.

I'll work on making this more random.